From 8879b685d38dfcc96a15ba926c12c1250276d350 Mon Sep 17 00:00:00 2001
From: Razvan Stefanescu <razvan.stefanescu@nxp.com>
Date: Fri, 5 Jul 2019 17:27:13 +0300
Subject: [PATCH 052/245] staging: fsl-dpaa2/ethsw: Remove netdevice on port
 probing error

commit 7162110cae171a1552b3d1bf461ebe02a9a4269f from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux.git

If the ethsw_port_init() call failed, the netdevice remains registered in
the system.

Use labels to ensure that netdevice is unregistered and freed in this case.

Signed-off-by: Razvan Stefanescu <razvan.stefanescu@nxp.com>
Signed-off-by: Ioana Ciornei <ioana.ciornei@nxp.com>
Link: https://lore.kernel.org/r/1562336836-17119-4-git-send-email-ioana.ciornei@nxp.com
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/staging/fsl-dpaa2/ethsw/ethsw.c | 16 +++++++++++++---
 1 file changed, 13 insertions(+), 3 deletions(-)

diff --git a/drivers/staging/fsl-dpaa2/ethsw/ethsw.c b/drivers/staging/fsl-dpaa2/ethsw/ethsw.c
index 146c0332e84b..fa59dbde562c 100644
--- a/drivers/staging/fsl-dpaa2/ethsw/ethsw.c
+++ b/drivers/staging/fsl-dpaa2/ethsw/ethsw.c
@@ -1390,13 +1390,23 @@ static int ethsw_probe_port(struct ethsw_core *ethsw, u16 port_idx)
 	err = register_netdev(port_netdev);
 	if (err < 0) {
 		dev_err(dev, "register_netdev error %d\n", err);
-		free_netdev(port_netdev);
-		return err;
+		goto err_register_netdev;
 	}
 
 	ethsw->ports[port_idx] = port_priv;
 
-	return ethsw_port_init(port_priv, port_idx);
+        err = ethsw_port_init(port_priv, port_idx);
+        if (err)
+                goto err_ethsw_port_init;
+
+        return 0;
+
+err_ethsw_port_init:
+        unregister_netdev(port_netdev);
+err_register_netdev:
+        free_netdev(port_netdev);
+
+        return err;
 }
 
 static int ethsw_probe(struct fsl_mc_device *sw_dev)
-- 
2.17.1

